src/uu/test/src/test.rs

```diff
@@ -163,6 +163,9 @@ fn eval(stack: &mut Vec<Symbol>) -> ParseResult<bool> {
         Some(Symbol::None) | None => Ok(false),
         Some(Symbol::BoolOp(op)) => {
             let b = eval(stack)?;
+            if stack.is_empty() {
+                return Err(ParseError::UnaryOperatorExpected(op.quote().to_string()));
+            }
```

Comment:
This works on the top-level, yes, but does not work if there are preceding expressions, e.g. `test foo -a ( -o bar )`.

In fact, this cannot work, since the information is lost after `parse()` is finished, and your code raises the error only later in `eval`. See #6203 for more examples.

Therefore, the error handling must happen during `parse()`, and doing it in `eval()` would only make it even harder to get it right.

Reply:
Woah, thank you for the detailed explanation in that issue. I'll take a look at how the parsing is done to see what's simple to change and watch the other issue for updates.

Reply:
Wanted to leave an update on this: I looked into using `nom` and `pest` for an easier to update grammar, and starting implementing the grammar in pest, but this was an awkward approach. `test` operates on a vector, while parsers usually operate on strings or something tokenized. Still trying things out, but wanted to relay I'm still interested in this